### PR TITLE
Refactor CA server

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
Please provide a description for what this PR is for.

cleanup some urgly code and also resolves a potential conflict

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[X] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any changes that may affect Istio users.

### CODE REVIEW

**Comment by howardjohn** [comment]
cc @shankgan @istio/wg-security-maintainers

File: security/pkg/pki/ca/ca.go [hunk 1]
```diff
@@ -81,45 +78,22 @@ type IstioCAOptions struct {
...
 	RotatorConfig *SelfSignedCARootCertRotatorConfig
 }
 
 // NewSelfSignedIstioCAOptions returns a new IstioCAOptions instance using self-signed certificate.
-func NewSelfSignedIstioCAOptions(ctx context.Context,
-	rootCertGracePeriodPercentile int, caCertTTL, rootCertCheckInverval, defaultCertTTL,
+func NewSelfSignedIstioCAOptions(
+	rootCertGracePeriodPercentile int, caCertTTL, rootCertCheckInterval, defaultCertTTL,
 	maxCertTTL time.Duration, org string, dualUse bool, namespace string,
-	readCertRetryInterval time.Duration, client corev1.CoreV1Interface,
```
Comment (shankgan) [inline security/pkg/pki/ca/ca.go:95]: Why was this removed? Is there no scenario where Istiod needs to wait for the caCerts to be loaded any longer?

Comment (hzxuzhonghu) [reply security/pkg/pki/ca/ca.go:95]: Currently no, because we donot have the wait until logic anymore

**Comment by hzxuzhonghu** [comment]
ping @istio/wg-security-maintainers

**Comment by lei-tang** [comment]
Can you provide details of 1) the improvements the refactoring makes, 2) the conflict the refactoring resolves?

File: pilot/pkg/bootstrap/istio_ca.go [hunk 2]
```diff
@@ -147,37 +142,16 @@ var (
...
-	return true
+	return features.EnableCAServer
 }
 
 // RunCA will start the cert signing GRPC service on an existing server.
 // Protected by installer options: the CA will be started only if the JWT token in /var/run/secrets
 // is mounted. If it is missing - for example old versions of K8S that don't support such tokens -
 // we will not start the cert-signing server, since pods will have no way to authenticate.
 func (s *Server) RunCA(grpc *grpc.Server, ca caserver.CertificateAuthority, opts *caOptions) {
-	if !s.EnableCA() {
```
Comment (lei-tang) [inline pilot/pkg/bootstrap/istio_ca.go:171]: Why these two checks are removed?

File: pilot/pkg/bootstrap/istio_ca.go [hunk 3]
```diff
@@ -147,37 +142,16 @@ var (
 // to have a central consistent endpoint to get whether CA functionality is
 // enabled in istiod. EnableCA() is called in multiple places.
 func (s *Server) EnableCA() bool {
-	if !features.EnableCAServer {
-		return false
-	}
-	// Log if we're using self-signed certs without K8S, in debug mode
-	if s.kubeClient == nil {
```
Comment (lei-tang) [inline pilot/pkg/bootstrap/istio_ca.go:154]: Why is this segment of code removed?

File: pilot/pkg/bootstrap/istio_ca.go (hunk 2 above)
Comment (shankgan) [inline pilot/pkg/bootstrap/istio_ca.go:171]: Why is this check being removed?

Comment (shankgan) [reply pilot/pkg/bootstrap/istio_ca.go:171]: ignore. its being checked in previous function

File: pilot/pkg/bootstrap/istio_ca.go [hunk 4]
```diff
@@ -404,9 +374,10 @@ func (s *Server) createIstioCA(client corev1.CoreV1Interface, opts *caOptions) (
 	// ca.go saves or uses the secret, but also writes to the configmap "istio-security", under caTLSRootCert
 	// rootCertRotatorChan channel accepts signals to stop root cert rotator for
 	// self-signed CA.
-	rootCertRotatorChan := make(chan struct{})
-	// Start root cert rotator in a separate goroutine.
-	istioCA.Run(rootCertRotatorChan)
+	s.addStartFunc(func(stop <-chan struct{}) error {
+		go istioCA.Run(stop)
```
Comment (shankgan) [inline pilot/pkg/bootstrap/istio_ca.go:378]: not needed. Run starts its own go routine.

File: pilot/pkg/bootstrap/istio_ca.go (hunk 3 above)
Comment (hzxuzhonghu) [reply pilot/pkg/bootstrap/istio_ca.go:154]: This is already covered by `func (s *Server) createIstioCA(client corev1.CoreV1Interface, opts *caOptions) (*ca.IstioCA, error) {
`

File: pilot/pkg/bootstrap/istio_ca.go (hunk 4 above)
Comment (hzxuzhonghu) [reply pilot/pkg/bootstrap/istio_ca.go:378]: good catch

**Comment by hzxuzhonghu** [comment]
>Can you provide details of 1) the improvements the refactoring makes, 2) the conflict the refactoring resolves?

@lei-tang For 1, there were some bad codes, say `maybeCreateCA`. For 2 Take a look at `NewSelfSignedIstioCAOptions
`, if we have multi istiod instances started at the same time and sign the cert, and the create the same secret at the same time, then at least one would fail.

File: pilot/pkg/bootstrap/istio_ca.go (hunk 3 above)
Comment (lei-tang) [reply pilot/pkg/bootstrap/istio_ca.go:154]: I did not find this segement of code in createIstioCA().
		// Without K8S the user needs to have the private key in a local file.
		// If that is missing - we'll generate an in-memory root for testing, and warn.
		signingKeyFile := path.Join(LocalCertDir.Get(), "ca-key.pem")
		if _, err := os.Stat(signingKeyFile); err != nil {
			log.Warnf("Will use in-memory root CA, no K8S access and no ca key file %s", signingKeyFile)
		}

File: pilot/pkg/bootstrap/istio_ca.go [hunk 5]
```diff
@@ -363,18 +336,15 @@ func (s *Server) createIstioCA(client corev1.CoreV1Interface, opts *caOptions) (
 	if _, err := os.Stat(signingKeyFile); err != nil && client != nil {
 		// The user-provided certs are missing - create a self-signed cert.
 		log.Info("Use self-signed certificate as the CA certificate")
-		// Abort after 20 minutes.
```
Comment (lei-tang) [inline pilot/pkg/bootstrap/istio_ca.go:366]: Why is the following code removed?
		// Abort after 20 minutes.
		ctx, cancel := context.WithTimeout(context.Background(), time.Minute*20)
		defer cancel()

File: pilot/pkg/bootstrap/istio_ca.go (hunk 3 above)
Comment (hzxuzhonghu) [reply pilot/pkg/bootstrap/istio_ca.go:154]: I mean this is duplicate with https://github.com/istio/istio/pull/30171/files/a611ae1cf6c2b87ff445a1d0a4c8300979441316#diff-0e039728d940bd528569ae3684ba2c8828df0f5d77456e3394fb0c94ea6e72e2L363

File: pilot/pkg/bootstrap/istio_ca.go (hunk 5 above)
Comment (hzxuzhonghu) [reply pilot/pkg/bootstrap/istio_ca.go:366]: In `NewSelfSignedIstioCAOptions`, the logic is:

1. get CA secret
2. if not found, sign root cert and create CA secret.
3. If CA secret create fails, then istiod will error exit. 

This flow is like most modules we have currently.

Comment (lei-tang) [reply pilot/pkg/bootstrap/istio_ca.go:366]: Thanks! Can you share a design document describing the changes the PR bring to the CA flows and present it in a Security Work Group meeting (https://docs.google.com/document/d/12Xz8fCuql2tPL-cpdLXGGE44kYT4o9NB6mljSiS_01U/edit#heading=h.o8pz6aqnzzgk)?

Comment (hzxuzhonghu) [reply pilot/pkg/bootstrap/istio_ca.go:366]: I donot think i can attend the meeting as the time is too late for my time zone.
